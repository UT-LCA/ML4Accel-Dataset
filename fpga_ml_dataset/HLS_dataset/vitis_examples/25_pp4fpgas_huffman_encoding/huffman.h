#include "ap_int.h"

// input number of symbols
const static int INPUT_SYMBOL_SIZE = 256;

// upper bound on codeword length during tree construction
const static int TREE_DEPTH = 64;

// maximum codeword tree length after rebalancing
const static int MAX_CODEWORD_LENGTH = 27;

// Should be log2(INPUT_SYMBOL_SIZE)
const static int SYMBOL_BITS = 10;

// Should be log2(TREE_DEPTH)
const static int TREE_DEPTH_BITS = 6;

// number of bits needed to record MAX_CODEWORD_LENGTH value
// Should be log2(MAX_CODEWORD_LENGTH)
const static int CODEWORD_LENGTH_BITS = 5;

// A marker for internal nodes
const static ap_uint<SYMBOL_BITS> INTERNAL_NODE = -1;

typedef ap_uint<MAX_CODEWORD_LENGTH> Codeword;
typedef ap_uint<MAX_CODEWORD_LENGTH + CODEWORD_LENGTH_BITS> PackedCodewordAndLength;
typedef ap_uint<CODEWORD_LENGTH_BITS> CodewordLength;
typedef ap_uint<32> Frequency;

struct Symbol {
	 ap_uint<SYMBOL_BITS> value;
	 ap_uint<32> frequency;
};

void huffman_encoding (
	Symbol in[INPUT_SYMBOL_SIZE],
	PackedCodewordAndLength encoding[INPUT_SYMBOL_SIZE],
	int *num_nonzero_symbols
);






void filter(Symbol in[INPUT_SYMBOL_SIZE],
            Symbol out[INPUT_SYMBOL_SIZE],
            int *num_symbols);
void sort(Symbol in[INPUT_SYMBOL_SIZE],
          int num_symbols,
          Symbol out[INPUT_SYMBOL_SIZE]);
void create_tree (
    Symbol in[INPUT_SYMBOL_SIZE],
    int num_symbols,
    ap_uint<SYMBOL_BITS> parent[INPUT_SYMBOL_SIZE-1],
    ap_uint<SYMBOL_BITS> left[INPUT_SYMBOL_SIZE-1],
    ap_uint<SYMBOL_BITS> right[INPUT_SYMBOL_SIZE-1]);

void compute_bit_length (
  ap_uint<SYMBOL_BITS> parent[INPUT_SYMBOL_SIZE-1],
  ap_uint<SYMBOL_BITS> left[INPUT_SYMBOL_SIZE-1],
  ap_uint<SYMBOL_BITS> right[INPUT_SYMBOL_SIZE-1],
  int num_symbols,
  ap_uint<SYMBOL_BITS> bit_length[TREE_DEPTH]);

void truncate_tree(
	ap_uint<SYMBOL_BITS> input_bit_length[TREE_DEPTH],
	ap_uint<SYMBOL_BITS> truncated_bit_length1[TREE_DEPTH],
	ap_uint<SYMBOL_BITS> truncated_bit_length2[TREE_DEPTH]);

void canonize_tree(
	Symbol sorted[INPUT_SYMBOL_SIZE],
	ap_uint<SYMBOL_BITS> num_symbols,
	ap_uint<SYMBOL_BITS> bit_length[TREE_DEPTH],
	CodewordLength symbol_bits[INPUT_SYMBOL_SIZE]);

void create_codeword(
  CodewordLength symbol_bits[INPUT_SYMBOL_SIZE],
  ap_uint<SYMBOL_BITS> bit_length[TREE_DEPTH],
  PackedCodewordAndLength encoding[INPUT_SYMBOL_SIZE]);


static unsigned int bit_reverse32(unsigned int input) {
	int i, rev = 0;
	for (i = 0; i < 32; i++) {
		rev = (rev << 1) | (input & 1);
		input = input >> 1;
	}
	return rev;
}
