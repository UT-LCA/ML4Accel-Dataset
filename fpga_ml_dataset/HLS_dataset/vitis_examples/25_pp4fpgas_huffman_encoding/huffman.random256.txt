0033
482d
18c8
678e
4aeb
3d76
2ce0
72b8
695c
5f9a
1653
6dfb
5afb
41c5
26f3
01f5
0bbd
2eb0
12e5
1546
7e91
3916
0f48
00a3
012e
3068
4417
4926
4d10
4dc1
1551
54e8
39bd
2d1c
0757
4dd2
644d
66c5
4295
26b0
7029
5d0d
7a64
7687
4513
1242
3b2f
1e29
6e67
1ade
63d5
6c06
7fa0
7fff
4e4f
3245
221d
2617
6b93
0314
3026
0be5
56b8
073c
012a
75a4
235a
22f8
4b4a
5882
6b40
5d07
3e1c
1a53
5f3c
3c00
3aa8
7987
5f53
0de6
4cb7
3159
5e1e
4dfc
494e
2e4a
1370
1cda
3675
66ce
423a
7ec1
603c
2c45
15ab
542c
3f00
082c
599b
40a7
12eb
7995
1229
73e4
58ba
26d4
36a3
090c
7bc3
577c
13a7
7053
6936
4a8a
1888
16cf
68a3
3cdf
13f3
408a
5dbc
33f4
23d3
48d6
575d
60c9
5c71
3ce0
0fc9
2f1e
6ae0
0488
4237
54e6
3698
0d70
798d
75f9
4661
2c53
3c6b
3009
6c73
2899
3a6b
22d7
7ddb
2628
5ea7
48a6
1920
617c
6b7c
32f0
4027
71fa
038e
7f59
4954
0681
440c
18e1
6bf2
5043
5436
195d
6bd5
0fd3
0e1c
5f28
283d
787e
24a8
2b16
11fe
5ddf
6ade
5aa9
4cde
5fae
2063
1288
003f
07d9
673c
6d2c
1afe
0ed6
46d9
01dd
0e9a
3a37
6052
57dd
4599
097f
37f0
19e3
5927
2534
37ef
1dca
4a01
4435
5082
148b
4091
7b4e
5918
7669
185a
2b0a
16de
7f6b
3a97
7fc8
0c85
500f
0c1f
3811
7745
063d
728c
2529
1d22
627a
349c
19e4
506e
4d5e
39d8
3bbb
4c8f
5148
6d73
6a1f
5002
5c50
4874
300e
17a0
5e7d
4718
73e3
1f20
1839
4d71
5972
4ade
2d01
3f54
0a54
5eda
4e61
4f72
5880
