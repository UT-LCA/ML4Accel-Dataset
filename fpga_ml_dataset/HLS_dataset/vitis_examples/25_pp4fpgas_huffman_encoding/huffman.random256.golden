0, ffff0
1, bc8
2, 8e9
3, 7
4, 1bc8
5, 7c8
6, 28e9
7, 807
8, 407
9, c07
10, 3dea
11, 207
12, 17c8
13, fc8
14, 18e9
15, ffed
16, 2feb
17, 38e9
18, 7dea
19, 3ea
20, a07
21, 1fc8
22, 43ea
23, 7ffef
24, 1ffee
25, 28
26, 1028
27, 828
28, 1828
29, 428
30, 23ea
31, 1428
32, c28
33, 4e9
34, afeb
35, 1c28
36, 607
37, e07
38, 228
39, 24e9
40, 107
41, 1228
42, 907
43, 507
44, a28
45, 63ea
46, 1a28
47, 14e9
48, d07
49, 34e9
50, 307
51, b07
52, 707
53, f07
54, 628
55, 1628
56, ce9
57, 2ce9
58, 87
59, 2ffed
60, e28
61, 13ea
62, 1e28
63, 6feb
64, 5ffee
65, 887
66, 1ce9
67, 3ce9
68, 128
69, 1128
70, 487
71, 928
72, 1928
73, 2e9
74, c87
75, 528
76, 1528
77, 287
78, a87
79, 53ea
80, d28
81, 1d28
82, 328
83, 1328
84, b28
85, 22e9
86, 33ea
87, 12e9
88, 1b28
89, 687
90, 728
91, e87
92, 187
93, 32e9
94, 73ea
95, 1728
96, f28
97, efeb
98, 1f28
99, a8
100, bea
101, 987
102, 4bea
103, 587
104, 10a8
105, ae9
106, 8a8
107, 1feb
108, d87
109, 18a8
110, 2bea
111, 387
112, b87
113, 4a8
114, 2ae9
115, 6bea
116, 787
117, 14a8
118, 1bea
119, ca8
120, 1ca8
121, 2a8
122, 1ae9
123, 12a8
124, aa8
125, f87
126, 1aa8
127, 6a8
128, 5bea
129, 3ae9
130, 47
131, 7fec
132, 16a8
133, ea8
134, 1ea8
135, 3bea
136, 847
137, 447
138, 1a8
139, 6e9
140, 11a8
141, 9a8
142, c47
143, 26e9
144, 19a8
145, 16e9
146, 247
147, 36e9
148, 5a8
149, 15a8
150, ee9
151, a47
152, 647
153, da8
154, 1da8
155, e47
156, 17fec
157, 147
158, 3a8
159, 9feb
160, 13a8
161, 2ee9
162, 947
163, ba8
164, 1ba8
165, 1ee9
166, 547
167, 7bea
168, 7ea
169, d47
170, 3ee9
171, 347
172, 1e9
173, 21e9
174, 47ea
175, 7a8
176, b47
177, 17a8
178, fa8
179, 747
180, 11e9
181, 27ea
182, 1ffff0
183, 5feb
184, f47
185, c7
186, 31e9
187, 67ea
188, 1fa8
189, 3ffee
190, 17ea
191, 68
192, 8c7
193, 1068
194, 868
195, dfeb
196, 1868
197, 9e9
198, 468
199, 29e9
200, 1468
201, 19e9
202, c68
203, 1c68
204, 268
205, 57ea
206, 1268
207, 4c7
208, a68
209, cc7
210, 39e9
211, 5e9
212, 37ea
213, 2c7
214, 1a68
215, ac7
216, 77ea
217, 668
218, fea
219, 1668
220, 6c7
221, 3feb
222, ec7
223, 25e9
224, 15e9
225, 1c7
226, e68
227, 35e9
228, 1e68
229, 168
230, 1168
231, 968
232, 1968
233, 568
234, 9c7
235, 5c7
236, 1568
237, d68
238, 1d68
239, 368
240, 4fea
241, 1368
242, b68
243, dc7
244, de9
245, 2de9
246, 1b68
247, 768
248, 1768
249, 1de9
250, f68
251, bfeb
252, 3c7
253, 1f68
254, e8
255, 10e8
